### PR TITLE
feat(vault config): set up access for users to access KVs through OIDC.

### PR DESCRIPTION
**Pull request by justbert**
Okay, so here's my PR for daaas/#23. 

It adds the ability to create or update the entity that links to an entity-alias.

### CODE REVIEW

File: vault.go [hunk 1]
```diff
@@ -118,7 +118,7 @@ func doMount(vc *vault.Client, name string) (string, error) {
 	return mountName, nil
 }
 
-func doKubernetesBackendRole(vc *vault.Client, authPath, name, roleName string, policies []string) error {
+func doKubernetesBackendRole(vc *vault.Client, authPath, namespace, roleName string, policies []string) error {
```
Comment (zachomedia) [inline vault.go]: Can we rename `namesapce` to `profileName`, since that's technically what it is.

File: vault.go [hunk 2]
```diff
@@ -171,7 +171,128 @@ func doMinioRole(vc *vault.Client, authPath, name string) error {
 	return nil
 }
 
-func doVaultConfiguration(vc *vault.Client, profileName string, minioInstances []string, kubernetesAuthPath string) error {
+func contains(s []string, e string) bool {
```
Comment (zachomedia) [inline vault.go]: There is already a helper for this: https://github.com/StatCan/kubeflow-controller/blob/master/utils.go#L4

**Review by zachomedia — CHANGES_REQUESTED** [review]
A few minor things but this looks good!

File: vault.go (hunk 1 above)
Comment (justbert) [reply vault.go]: Absolutely!

Comment (justbert) [reply vault.go]: Absolutely!

File: vault.go (hunk 2 above)
Comment (justbert) [reply vault.go]: Ouuuu! Thanks!

File: controller.go [hunk 3]
```diff
@@ -473,7 +468,8 @@ func (c *Controller) syncHandler(key string) error {
 	}
 
 	// Configure vault
-	err = doVaultConfiguration(c.vaultClient, profile.Name, c.minioInstances, c.kubernetesAuthPath)
+	err = c.vaultConfigurer.ConfigVaultForProfile(profile.Name, profile.Spec.Owner.Name, []string{})
+	//doVaultConfiguration(c.vaultClient, profile.Name, profile.Spec.Owner.Name, c.minioInstances, c.kubernetesAuthPath, c.oidcAuthAccessor)
```
Comment (zachomedia) [inline controller.go]: Please remove the old function call.

File: vault.go [hunk 4]
```diff
@@ -30,54 +104,51 @@ path "{{ $instance }}/keys/{{ $profile }}" {
...
-func cleanName(name string) string {
-	return strings.ReplaceAll(name, "_", "-")
+	return w.String(), nil
 }
 
-func doPolicy(vc *vault.Client, name, mountName string, minioInstances []string) (string, error) {
-	policyName, policy := generatePolicy(name, mountName, minioInstances)
+// Writes a policy to Vault
+func (vc *VaultConfigurerStruct) doPolicy(name string, mountName string) (string, error) {
+	policy, err := generatePolicy(name, vc.MinioInstances)
```
Comment (zachomedia) [inline vault.go:127]: Does this return an error if the policy doesn't exist? Just want to make sure it won't ever not create policies.

File: vault.go [hunk 5]
```diff
@@ -212,12 +466,38 @@ func doVaultConfiguration(vc *vault.Client, profileName string, minioInstances [
 	//
 	// Add MinIO role
 	//
-	for _, instance := range minioInstances {
-		if err := doMinioRole(vc, instance, name); err != nil {
+	//for _, instance := range vc.MinioInstances {
```
Comment (zachomedia) [inline vault.go]: Is there a reason this is commented out?

**Review by zachomedia — CHANGES_REQUESTED** [review]
@justbert Thanks for the updates to this, I took an initial look through and left some comments beind - but doesn't seem like anyhting major. Great work!

File: vault.go (hunk 4 above)
Comment (justbert) [reply vault.go:127]: the generatePolicy()? I added error handling if the templating through an error.

File: vault.go (hunk 5 above)
Comment (justbert) [reply vault.go]: Fat fingers!

File: controller.go [hunk 6]
```diff
@@ -473,7 +468,7 @@ func (c *Controller) syncHandler(key string) error {
 	}
 
 	// Configure vault
-	err = doVaultConfiguration(c.vaultClient, profile.Name, c.minioInstances, c.kubernetesAuthPath)
+	err = c.vaultConfigurer.ConfigVaultForProfile(profile.Name, profile.Spec.Owner.Name, []string{})
```
Comment (zachomedia) [inline controller.go]: Is there a reason this is a hard coded empty array instead of a list of users from the namespace?

File: defaults.go [hunk 7]
```diff
@@ -17,6 +18,10 @@ var (
 	PodDefaults = make(map[string]NewPodDefaultFunc)
 )
 
+func cleanName(name string) string {
```
Comment (zachomedia) [inline defaults.go]: Can we add this to utils.go instead?

File: main.go [hunk 8]
```diff
@@ -90,13 +95,20 @@ func main() {
 		klog.Fatalf("Error initializing Vault client: %s", err)
 	}
 
-	controller := NewController(kubeClient, kubeflowClient,
+	var vaultConfigurer VaultConfigurer
+	vaultConfigurer = NewVaultConfigurer(vc,
```
Comment (zachomedia) [inline main.go]: You can do:

```go
vaultConfigurer := NewVaultConfigurer
```

And get rid of the var line from above it.

**Review by zachomedia — CHANGES_REQUESTED** [review]
A couple of additional changes included.

**Comment by justbert** [comment]
Goes with [PR#15](https://github.com/StatCan/terraform-vault-daaas/pull/15).

**Review by zachomedia — APPROVED** [review]
This set of changes LGTM. I don't have an easy way to test this all, so we'll have to do that when it's deployed.

**Comment by zachomedia** [comment]
@justbert Can you rebase this against the latest master so that it removes the merge conflict?

**Comment by justbert** [comment]
@zachomedia Done!

**Comment by justbert** [comment]
@zachomedia Is this ready to be merged? Do we need to wait for the evening?

**Comment by zachomedia** [comment]
@justbert I wanted to wait until you were available in case there are issues. Maybe this afternoon we can do it.

**Pull request merged**